server/athenian/api/internal/team_sync.py

```diff
@@ -0,0 +1,325 @@
+        meta_orgs_query = sa.select(MetadataOrganization.id).where(
+            MetadataOrganization.acc_id.in_(meta_ids),
+        )
+        columns = (MetadataTeam.name, MetadataTeam.id, MetadataTeam.parent_team_id)
+        query = sa.select(columns).where(
+            MetadataTeam.acc_id.in_(meta_ids), MetadataTeam.organization_id.in_(meta_orgs_query),
+        )
+        return cls(await mdb.fetch_all(query), meta_ids, mdb)
+
+    def get_new(self, state_teams: _StateTeams) -> Sequence[Row]:
+        existing = state_teams.existing_by_origin_node_id
+        return [r for r in self._rows if r[MetadataTeam.id.name] not in existing]
+
+    def get_existing(self, state_teams: _StateTeams) -> Sequence[Row]:
+        existing = state_teams.existing_by_origin_node_id
+        return [r for r in self._rows if r[MetadataTeam.id.name] in existing]
+
+    async def get_all_members(self) -> dict[int, list[int]]:
+        ids = [r[MetadataTeam.id.name] for r in self._rows]
+        return await get_meta_teams_members(ids, self._meta_ids, self._mdb)
+
+    @property
+    def by_id(self) -> dict[int, Row]:
+        return self._by_id
+
+
+class _StateTeams:
+    def __init__(self, rows: Sequence[Row], root_team_id: int):
+        self._rows = rows
+        self._root_team_id = root_team_id
+        self._existing = {id_: r for r in rows if (id_ := r[Team.origin_node_id.name]) is not None}
+        self._created: dict[int, int] = {}
+        self._check_unmapped()
+
+    @classmethod
+    async def from_db(cls, account: int, sdb_conn: Connection) -> _StateTeams:
+        where = [Team.owner_id == account, Team.name != Team.BOTS, Team.parent_id.isnot(None)]
+        query = sa.select(Team).where(*where)
+        root_team_id = (await get_root_team(account, sdb_conn))[Team.id.name]
+        rows = await sdb_conn.fetch_all(query)
+        return cls(rows, root_team_id)
+
+    @property
+    def existing_by_origin_node_id(self) -> dict[int, Row]:
+        return self._existing
+
+    @property
+    def root_team_id(self) -> int:
+        return self._root_team_id
+
+    def track_created_team(self, id_: int, origin_node_id) -> None:
+        self._created[origin_node_id] = id_
+
+    def get_id(self, origin_node_id: int) -> int:
+        existing_row = self._existing.get(origin_node_id)
+        if existing_row is not None:
+            return existing_row[Team.id.name]
+        return self._created[origin_node_id]
+
+    def get_gone(self, meta_teams) -> Sequence[Row]:
+        rows = self._existing.values()
+        return [r for r in rows if r[Team.origin_node_id.name] not in meta_teams.by_id]
+
+    def _check_unmapped(self) -> None:
+        unmapped = [r[Team.id.name] for r in self._rows if r[Team.origin_node_id.name] is None]
+        if unmapped:
+            raise SyncTeamsError(
+                f"Cannot sync teams, account has some teams not mapped to mdb: {unmapped}",
+            )
+
+
+class _NameMangler:
+    _TMP_PREFIX = "$32$àē_ì_ōū$32$"
```

Comment:
Github team names may not contain "$" so it's already fine to prefix with one symbol. Source: https://github.com/dead-claudia/github-limits#team-names